### ISSUE TITLE
Set-Location struggles with UNC paths

### ISSUE THREAD

**Issue by 9Rune5** [PowerShell/PowerShell#20879]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

This closed issue still reproduces: https://github.com/PowerShell/PowerShell/issues/13854

Set-Location sometimes struggles with UNC paths.

### Expected behavior

```console
`Set-Location \\stallbacka\home\config

Set-Location \
# this should set the UNC root as the new working directory`
```

### Actual behavior

```console
`Set-Location \\stallbacka\home\config
Set-Location \
#Set-Location: Cannot find path 'Microsoft.PowerShell.Core\FileSystem::' because it does not exist.`

The same error can be triggered by going to e.g. cert: first:
`Set-Location Cert:\
Set-Location \\stallbacka\home\config
#Set-Location: Cannot find path 'Microsoft.PowerShell.Core\FileSystem::' because it does not exist.`
```

### Error details

```console
Exception             :
    Type                 : System.Management.Automation.ItemNotFoundException
    ErrorRecord          :
        Exception             :
            Type    : System.Management.Automation.ParentContainsErrorRecordException
            Message : Cannot find path 'Microsoft.PowerShell.Core\FileSystem::' because it does not exist.
            HResult : -2146233087
        TargetObject          : Microsoft.PowerShell.Core\FileSystem::
        CategoryInfo          : ObjectNotFound: (Microsoft.PowerShell.Core\FileSystem:::String) [],
ParentContainsErrorRecordException
        FullyQualifiedErrorId : PathNotFound
    ItemName             : Microsoft.PowerShell.Core\FileSystem::
    SessionStateCategory : Drive
    TargetSite           :
        Name          : ExpandMshGlobPath
        DeclaringType : System.Management.Automation.LocationGlobber, System.Management.Automation, Version=7.4.0.500,
Culture=neutral, PublicKeyToken=31bf3856ad364e35
        MemberType    : Method
        Module        : System.Management.Automation.dll
    Message              : Cannot find path 'Microsoft.PowerShell.Core\FileSystem::' because it does not exist.
    Source               : System.Management.Automation
    HResult              : -2146233087
    StackTrace           :
   at System.Management.Automation.LocationGlobber.ExpandMshGlobPath(String path, Boolean allowNonexistingPaths,
PSDriveInfo drive, ContainerCmdletProvider provider, CmdletProviderContext context)
   at System.Management.Automation.LocationGlobber.ResolveDriveQualifiedPath(String path, CmdletProviderContext
context, Boolean allowNonexistingPaths, CmdletProvider& providerInstance)
   at System.Management.Automation.LocationGlobber.GetGlobbedMonadPathsFromMonadPath(String path, Boolean
allowNonexistingPaths, CmdletProviderContext context, CmdletProvider& providerInstance)
   at System.Management.Automation.SessionStateInternal.SetLocation(String path, CmdletProviderContext context,
Boolean literalPath)
   at Microsoft.PowerShell.Commands.SetLocationCommand.ProcessRecord()
TargetObject          : Microsoft.PowerShell.Core\FileSystem::
CategoryInfo          : ObjectNotFound: (Microsoft.PowerShell.Core\FileSystem:::String) [Set-Location],
ItemNotFoundException
FullyQualifiedErrorId : PathNotFound,Microsoft.PowerShell.Commands.SetLocationCommand
InvocationInfo        :
    MyCommand        : Set-Location
    ScriptLineNumber : 1
    OffsetInLine     : 1
    HistoryId        : 9
    Line             : cd \
    Statement        : cd \
    PositionMessage  : At line:1 char:1
                       + cd \
                       + ~~~~
    InvocationName   : cd
    CommandOrigin    : Internal
ScriptStackTrace      : at <ScriptBlock>, <No file>: line 1
PipelineIterationInfo :
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0
PSEdition                      Core
GitCommitId                    7.4.0
OS                             Microsoft Windows 10.0.22631
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

![image](https://github.com/PowerShell/PowerShell/assets/9988080/e4c65cf6-112c-4dd6-8ed0-b52d5de06b4a)

**Comment by jhoneill** [comment]
TLDR -  It needs tidying up, and for more cases than this one. 

Elsewhere in Windows is not consistent about whether you can move to a UNC path or only to a drive mapped to one

<img width="513" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/b037d02f-fe39-4f6e-b8e3-ef00129bbf6b">

If you map a drive letter to a directory under a UNC path you can't cd up from there

<img width="476" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/43031def-0337-44f8-8370-b87fb414976a">

In explorer if I type \ from the `J:` drive or from `\\server\share\directory`  it takes me to the root of **C:**

But if cd to a unc subpath in PowerShell  you can cd to `..` but not to `\` should it go to the main file system like linux, and Windows Explorer or to the root of the share? It's poorly defined and instead of picking one the implementors have it error.  
 
<img width="499" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/c18738c1-03df-4416-a5d7-0519c5174cd9">

And as you say for other drives  `cd function:\` works , but then cd `\` fails 
<img width="474" alt="image" src="https://github.com/PowerShell/PowerShell/assets/16209593/2951603b-f064-4b4e-84d2-c2a66b958e9d">

Providers like the registry which have a hierarchy treat cd \ correctly, but the function etc don't (IMHO they should treat a cd \ as a No-op)

**Comment by mklement0** [comment]
@9Rune5:

* #13854 is undoubtedly still a _bug_.

  * The only sensible interpretation of `\` or `/` relative to a _UNC location_ is the _root of the share_,  namely `\\stallbacka\home` in your example.

  * As an aside: On _Unix_-like platforms, `/` and `\` are invariably interpreted as referring to the root of the _file-system_, _irrespective of the provider underlying the current location_. This means that in order to change to the root of the _specific provider_ that underlies the current location, you need to _explicitly use the current location's drive name_, e.g. `Set-Location Temp:\`

* As for trying  `Set-Location \\stallbacka\home\config` _while a provider OTHER than `FileSystem`_ underlies the current location, such as the `Certificate` provider, via its `Cert:\` drive:

  * The problem is that a path starting with `\\` currently isn't recognized _as a `FileSystem` provider location_ on Windows.

    * Currently,  the logic is to interpret a `\\` prefix as _drive-and-therefore-provider-relative_, presumably based on the logic that  _accidental duplication of path separators_ as  `\\` and `//` is acceptable in calls such as `Get-ChildItem C:\\` (on Windows) and `Get-ChildItem //tmp` (on Unix-like platforms).

   * However, arguably it would be more sensible to interpret  `\\` at the _start_ of a path _on Windows_ as unequivocally referring to a UNC _file-system_ location: see https://github.com/PowerShell/PowerShell/issues/10522#issuecomment-1233028415

**Comment by 9Rune5** [comment]
@mklement0 A random thought: Would it make things more consistent if UNC paths were handled by a UNC provider? I.e. `Set-Location unc:\\stallbacka\home`.

**Comment by mklement0** [comment]
@9Rune5, I don't think that's necessary - my suggestion is to simply start interpreting UNC paths _unconditionally_ as such; while _technically_ a breaking change, I would expect very little, if any, existing code to be impacted; see:

* #20897
